File: backend/app/models/persona.py
"""Persona domain model: an evidence-grounded synthetic user persona."""

import enum
from datetime import UTC, datetime
from typing import TYPE_CHECKING

from sqlalchemy import JSON, DateTime, Enum, ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database.base import Base

if TYPE_CHECKING:
    from app.models.project import Project


class ConfidenceLevel(enum.StrEnum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


def _utcnow() -> datetime:
    return datetime.now(UTC)


class Persona(Base):
    __tablename__ = "personas"

    id: Mapped[int] = mapped_column(primary_key=True)
    project_id: Mapped[int] = mapped_column(
        ForeignKey("projects.id", ondelete="CASCADE"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    segment_label: Mapped[str] = mapped_column(String(255), nullable=False)
    summary: Mapped[str] = mapped_column(Text, nullable=False)
    goals: Mapped[list[str]] = mapped_column(JSON, nullable=False, default=list)
    pain_points: Mapped[list[str]] = mapped_column(JSON, nullable=False, default=list)
    constraints: Mapped[list[str]] = mapped_column(JSON, nullable=False, default=list)
    behaviors: Mapped[list[str]] = mapped_column(JSON, nullable=False, default=list)
    evidence_references: Mapped[list[dict]] = mapped_column(JSON, nullable=False, default=list)
    unsupported_assumptions: Mapped[list[str]] = mapped_column(JSON, nullable=False, default=list)
    confidence_level: Mapped[ConfidenceLevel] = mapped_column(
        Enum(ConfidenceLevel, native_enum=False, length=10), nullable=False
    )
    prompt_version: Mapped[str] = mapped_column(String(50), nullable=False)
    model_name: Mapped[str] = mapped_column(String(100), nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=_utcnow
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=_utcnow, onupdate=_utcnow
    )

    project: Mapped["Project"] = relationship(back_populates="personas")

File: frontend/types/persona.ts
export type ConfidenceLevel = "low" | "medium" | "high";

export interface EvidenceReference {
  evidence_item_id: number;
  supported_claims: string[];
}

export interface Persona {
  id: number;
  project_id: number;
  name: string;
  segment_label: string;
  summary: string;
  goals: string[];
  pain_points: string[];
  constraints: string[];
  behaviors: string[];
  evidence_references: EvidenceReference[];
  unsupported_assumptions: string[];
  confidence_level: ConfidenceLevel;
  prompt_version: string;
  model_name: string;
  created_at: string;
  updated_at: string;
}

export interface PersonaGenerateInput {
  persona_count: number;
  selected_evidence_ids?: number[] | null;
  focus?: string | null;
}

export interface PersonaGenerateResponse {
  project_id: number;
  prompt_version: string;
  model_name: string;
  persona_count: number;
  personas: Persona[];
}
